Python/specialize.c

```diff
@@ -362,94 +427,134 @@ _Py_Specialize_LoadAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name, Sp
-        // We found an attribute with a data-like descriptor.
-        PyTypeObject *dtype = Py_TYPE(descr);
-        if (dtype != &PyMemberDescr_Type) {
-            SPECIALIZATION_FAIL(LOAD_ATTR, Py_TYPE(owner), name, "not a member descriptor");
+    PyObject *descr;
+    DesciptorClassification kind = analyze_descriptor(type, name, &descr);
+    switch(kind) {
+        case OVERRIDING:
+            SPECIALIZATION_FAIL(LOAD_ATTR, type, name, "overriding descriptor");
             goto fail;
-        }
-        // It's a slot
-        PyMemberDescrObject *member = (PyMemberDescrObject *)descr;
-        struct PyMemberDef *dmem = member->d_member;
-        if (dmem->type != T_OBJECT_EX) {
-            SPECIALIZATION_FAIL(LOAD_ATTR, Py_TYPE(owner), name, "non-object slot");
+        case METHOD:
+            SPECIALIZATION_FAIL(LOAD_ATTR, type, name, "method");
             goto fail;
-        }
-        Py_ssize_t offset = dmem->offset;
-        if (offset != (uint16_t)offset) {
-            SPECIALIZATION_FAIL(LOAD_ATTR, Py_TYPE(owner), name, "offset out of range");
+        case PROPERTY:
+            SPECIALIZATION_FAIL(LOAD_ATTR, type, name, "property");
             goto fail;
+        case OBJECT_SLOT:
+        {
+            PyMemberDescrObject *member = (PyMemberDescrObject *)descr;
+            struct PyMemberDef *dmem = member->d_member;
+            Py_ssize_t offset = dmem->offset;
+            if (offset != (uint16_t)offset) {
+                SPECIALIZATION_FAIL(LOAD_ATTR, type, name, "offset out of range");
+                goto fail;
+            }
+            assert(dmem->type == T_OBJECT_EX);
+            assert(offset > 0);
+            cache0->index = (uint16_t)offset;
+            cache1->tp_version = type->tp_version_tag;
+            *instr = _Py_MAKECODEUNIT(LOAD_ATTR_SLOT, _Py_OPARG(*instr));
+            goto success;
         }
-        assert(offset > 0);
-        cache0->index = (uint16_t)offset;
-        cache1->tp_version = type->tp_version_tag;
-        *instr = _Py_MAKECODEUNIT(LOAD_ATTR_SLOT, _Py_OPARG(*instr));
-        goto success;
-    }
-    // No desciptor
-    if (type->tp_dictoffset <= 0) {
-        SPECIALIZATION_FAIL(LOAD_ATTR, Py_TYPE(owner), name, "no dict or negative offset");
-        goto fail;
-    }
-    PyObject **dictptr = (PyObject **) ((char *)owner + type->tp_dictoffset);
-    if (*dictptr == NULL || !PyDict_CheckExact(*dictptr)) {
-        SPECIALIZATION_FAIL(LOAD_ATTR, Py_TYPE(owner), name, "no dict or not a dict");
+        case __CLASS__:
+        {
+            Py_ssize_t offset = offsetof(PyObject, ob_type);
+            assert(offset == (uint16_t)offset);
+            cache0->index = (uint16_t)offset;
+            cache1->tp_version = type->tp_version_tag;
+            *instr = _Py_MAKECODEUNIT(LOAD_ATTR_SLOT, _Py_OPARG(*instr));
+            goto success;
+        }
+        case OTHER_SLOT:
+            SPECIALIZATION_FAIL(LOAD_ATTR, type, name, "non-object slot");
```

Comment:
This is certainly less common than "object slots".  Would there be enough value to specializing any other member slot types later?  Either way, would it make sense to identify at least some of them now in `analyze_descriptor()` rather than lumping them all into "OTHER_SLOT"?  (My guess is that they're not common enough to be worth identifying specifically.)

Reply:
It might well be worth add integer slots, and getset descriptors, and a bunch of other things.
But first, we really need data to decide on what is worthwhile.